design_documents/parametrizing_tests_with_fixed_values.md

```diff
@@ -0,0 +1,192 @@
+# Parametrizing Tests With Fixed Values
+
+## Context
+
+Various test runners for other programming languages provide ways of parametrizing test cases with fixed values.
+This way, multiple test cases can be executed for the same test source code.
+
+## Goal
+
+Propose a way of adding support for parametrized tests to Starknet Foundry's `snforge`.
+
+## Proposed Solution
+
+Introduce two new attributes: `parametrize` and `case`.
+
+- `parametrize` - defines a parametrized test.
+- `case` - defines a specific test case
+
+Generate code for each of the test cases so no handling of values injection, etc. is necessary.
+The compiler would then validate if the values for the generated code are actually valid.
+Thanks to that, we could have failures for invalid arguments early in the execution.
+
+### Parametrization With Simple Arguments
+
+```cairo
+#[parametrize]
+#[case(1, 3)]
+#[case(3, 5)]
+#[test]
+fn my_test(a: felt252, b: u32) {
+    // case a = 1, b = 3
+    // case a = 3, b = 5
+    // ...
+}
+```
+
+For test cases like this, we would generate two actual test cases:
+
+```cairo
+#[test]
+fn my_test_case_1() {
+  let a: felt252 = 3;
+  let b: u32 = 1;
+}
+
+#[test]
+fn my_test_case_1() {
+  let a: felt252 = 3;
+  let b: u32 = 5;
+}
+```
+
+The explicit type annotations for each argument are necessary.
```

Comment:
I think those are necessary by language semantics, nothing this doc-specific

Reply:
I think language semantics are quite important in the case of this document.

Reply:
@cptartur then you should add a section as well that functions should have names and are prefixed with `fn` keyword and functions can be called from other functions etc. You write a document under specific assumptions and don't mix existing mechanisms with your design otherwise it is hard to read.

Reply:
The point of this section is to ensure that generated code has this type annotations included, because it's important for its behavior.

I can remove this section if you really think it's not necessary, but it seems at list a bit useful.

Reply:
Agree with Maks, this section is just describing Cairo spec

Reply:
Not relevant anymore